src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs

```diff
@@ -0,0 +1,295 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using SFA.DAS.CollectionEarnings.DataLock.Application.DataLock;
+using SFA.DAS.CollectionEarnings.DataLock.Application.DataLock.Matcher;
+using SFA.DAS.CollectionEarnings.DataLock.Domain;
+using SFA.DAS.CollectionEarnings.DataLock.Domain.Extensions;
+using SFA.DAS.CollectionEarnings.DataLock.Infrastructure.Data.Entities;
+using SFA.DAS.Payments.DCFS.Domain;
+using SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.CollectionEarnings.DataLock.Services
+{
+    public interface IValidateDatalocks
+    {
+        DatalockValidationResult ValidateDatalockForProvider(
+            ProviderCommitments providerCommitments, 
+            IEnumerable<RawEarning> providerEarnings, 
+            ImmutableHashSet<long> accountsWithNonPayableFlagSet);
+    }
+
+    public class DatalockValidationService : IValidateDatalocks
+    {
+        private readonly IMatcher _datalockMatcher;
+
+        public DatalockValidationService(IMatcher datalockMatcher)
+        {
+            _datalockMatcher = datalockMatcher;
+        }
+
+        public DatalockValidationResult ValidateDatalockForProvider(
+            ProviderCommitments providerCommitments, 
+            IEnumerable<RawEarning> providerEarnings,
+            ImmutableHashSet<long> accountsWithNonPayableFlagSet)
+        {
+            var earningsByLearner = new ProviderEarnings(providerEarnings);
+            var processedLearners = new HashSet<long>();
+            
+            var result = new DatalockValidationResult();
+
+            foreach (var uln in providerCommitments.AllUlns())
+            {
+                processedLearners.Add(uln);
+
+                var learnerCommitments = providerCommitments.CommitmentsForLearner(uln);
+                var earnings = earningsByLearner.EarningsForLearner(uln);
+
+                foreach (var earning in earnings)
+                {
+                    if (earning.HasNonIncentiveEarnings())
+                    {
+                        ProcessEarning(accountsWithNonPayableFlagSet, earning, 
```

Comment:
What does ProcessEarning and why do we have to do it seperately for IncentivePayments?